File: python/src/cltf/observations.py
#!/usr/bin/env python3
"""
Script: observations.py
Objective: Prepare replicate-level herbicide observations for CLTF analysis.
Author: Yi Yu
Created: 2026-06-24
Last updated: 2026-06-24
Inputs: Herbicide workbooks or tidy concentration vectors.
Outputs: Depth intervals, non-detect fields, summaries, and applied mass.
Usage: Import public functions from cltf or cltf.observations.
Dependencies: math, pathlib, numpy, pandas, cltf.concentration
"""

from __future__ import annotations

import math
from pathlib import Path
from typing import Sequence

import numpy as np
import pandas as pd
from numpy.typing import ArrayLike

from .concentration import soil_mass_kg_ha


_DEPTH_INTERVALS = {
    "SA:10cm": (0.0, 100.0),
    "SA:30cm": (100.0, 300.0),
    "NSW:15cm": (0.0, 150.0),
    "NSW:30cm": (150.0, 300.0),
    "QLD:10cm": (0.0, 100.0),
    "QLD:30cm": (100.0, 300.0),
}

_SITE_IDS = {
    "SA": "SA_Minnipa",
    "NSW": "NSW_Griffith",
    "QLD": "QLD_Wellcamp",
}

_IDENTIFIER_COLUMNS = (
    "Soil",
    "Irrigation",
    "Timepoint",
    "Crop_2024",
    "Depth",
    "Sample_date",
)


def depth_interval_mm(sheet: str, depth_label: str) -> tuple[float, float]:
    """Map a workbook jurisdiction/depth label to an explicit interval."""

    normalized_depth = "".join(str(depth_label).strip().lower().split())
    key = f"{str(sheet).strip().upper()}:{normalized_depth}"
    try:
        return _DEPTH_INTERVALS[key]
    except KeyError as error:
        raise ValueError(
            f"Unsupported sheet/depth combination: {key}"
        ) from error


def prepare_non_detects(
    concentration_ug_kg: ArrayLike,
    is_non_detect: ArrayLike,
    detection_limit_ug_kg: ArrayLike,
) -> pd.DataFrame:
    """Prepare analysis concentrations with explicit detection-limit handling."""

    concentration = np.atleast_1d(
        np.asarray(concentration_ug_kg, dtype=float)
    )
    non_detect_raw = np.atleast_1d(np.asarray(is_non_detect, dtype=object))
    detection_limit = np.atleast_1d(
        np.asarray(detection_limit_ug_kg, dtype=float)
    )
    if len({len(concentration), len(non_detect_raw), len(detection_limit)}) != 1:
        raise ValueError("Non-detect input vectors must have equal lengths")
    if not np.all(np.isfinite(concentration)):
        raise ValueError("Reported concentrations must be finite")
    if pd.isna(non_detect_raw).any():
        raise ValueError("is_non_detect cannot contain missing values")

    non_detect = non_detect_raw.astype(bool)
    substituted = (
        non_detect
        & np.isfinite(detection_limit)
        & (detection_limit > 0)
    )
    excluded_zero = (concentration <= 0) & ~substituted
    analysis = concentration.copy()
    analysis[substituted] = detection_limit[substituted] / 2.0
    analysis[excluded_zero] = np.nan

    return pd.DataFrame(
        {
            "analysis_concentration_ug_kg": analysis,
            "lod_substituted": substituted,
            "excluded_zero": excluded_zero,
        }
    )


def geometric_concentration(
    data: pd.DataFrame,
    group_columns: Sequence[str],
) -> pd.DataFrame:
    """Calculate grouped geometric means and geometric standard deviations."""

    concentration_column = "analysis_concentration_ug_kg"
    if concentration_column not in data.columns:
        raise ValueError(
            f"data must contain {concentration_column}"
        )
    groups = list(group_columns)
    if not groups or not set(groups).issubset(data.columns):
        raise ValueError("group_columns must identify columns in data")

    rows: list[dict[str, object]] = []
    grouper: str | list[str] = groups[0] if len(groups) == 1 else groups
    grouped = data.groupby(grouper, sort=True, dropna=True, observed=True)
    for key, group in grouped:
        keys = key if isinstance(key, tuple) else (key,)
        values = pd.to_numeric(
            group[concentration_column],
            errors="coerce",
        ).to_numpy(dtype=float)
        values = values[np.isfinite(values) & (values > 0)]
        log_values = np.log(values)
        row = dict(zip(groups, keys))
        row["n"] = len(values)
        row["geometric_mean_ug_kg"] = (
            float(np.exp(np.mean(log_values))) if len(values) else np.nan
        )
        row["geometric_sd"] = (
            float(np.exp(np.std(log_values, ddof=1)))
            if len(values) > 1
            else np.nan
        )
        rows.append(row)

    return pd.DataFrame(
        rows,
        columns=groups + ["n", "geometric_mean_ug_kg", "geometric_sd"],
    )


def _normalize_sample_date(values: pd.Series) -> pd.Series:
    if pd.api.types.is_numeric_dtype(values):
        return pd.to_datetime(
            values,
            unit="D",
            origin="1899-12-30",
            errors="coerce",
        ).dt.normalize()
    return pd.to_datetime(values, errors="coerce").dt.normalize()


def _text_column(
    data: pd.DataFrame,
    column: str,
    default: object,
) -> pd.Series:
    values = (
        data[column]
        if column in data.columns
        else pd.Series(default, index=data.index)
    )
    return values.astype("string").str.strip()


def _prepare_workbook_sheet(path: Path, sheet: str) -> pd.DataFrame:
    raw = pd.read_excel(path, sheet_name=sheet)
    missing = {
        "Soil",
        "Timepoint",
        "Depth",
        "Sample_date",
    }.difference(raw.columns)
    if missing:
        raise ValueError(
            f"Sheet {sheet} is missing required columns: {sorted(missing)}"
        )

    sheet_key = sheet.upper()
    try:
        site_id = _SITE_IDS[sheet_key]
    except KeyError as error:
        raise ValueError(f"Unsupported workbook sheet: {sheet}") from error

    herbicide_columns = [
        column
        for column in raw.columns
        if column not in _IDENTIFIER_COLUMNS
    ]
    frames: list[pd.DataFrame] = []
    for herbicide in herbicide_columns:
        concentration = pd.to_numeric(raw[herbicide], errors="coerce")
        keep = np.isfinite(concentration.to_numpy(dtype=float))
        if not np.any(keep):
            continue

        selected = raw.loc[keep]
        frames.append(
            pd.DataFrame(
                {
                    "site_id": site_id,
                    "source_sheet": sheet,
                    "source_row": selected.index.to_numpy(dtype=int) + 2,
                    "soil_group": _text_column(raw, "Soil", pd.NA).loc[keep],
                    "treatment": _text_column(
                        raw,
                        "Irrigation",
                        "All",
                    ).loc[keep],
                    "crop_2024": _text_column(
                        raw,
                        "Crop_2024",
                        pd.NA,
                    ).loc[keep],
                    "timepoint": _text_column(
                        raw,
                        "Timepoint",
                        pd.NA,
                    ).loc[keep],
                    "depth_label": _text_column(
                        raw,
                        "Depth",
                        pd.NA,
                    ).loc[keep],
                    "sample_date": _normalize_sample_date(
                        raw["Sample_date"]
                    ).loc[keep],
                    "herbicide": str(herbicide),
                    "concentration_ug_kg": concentration.loc[keep],
                }
            ).reset_index(drop=True)
        )

    if not frames:
        raise ValueError(f"No herbicide concentrations found in sheet {sheet}")
    return pd.concat(frames, ignore_index=True)


def read_herbicide_workbook(
    path: str | Path,
    sheets: Sequence[str] = ("SA", "NSW", "Qld"),
) -> pd.DataFrame:
    """Read and normalize the herbicide dissipation workbook."""

    workbook_path = Path(path)
    if not workbook_path.exists():
        raise FileNotFoundError(
            f"Observation workbook does not exist: {workbook_path}"
        )

    observations = pd.concat(
        [
            _prepare_workbook_sheet(workbook_path, sheet)
            for sheet in sheets
        ],
        ignore_index=True,
    )
    intervals = [
        depth_interval_mm(sheet, depth)
        for sheet, depth in zip(
            observations["source_sheet"],
            observations["depth_label"],
        )
    ]
    observations["depth_top_mm"] = [interval[0] for interval in intervals]
    observations["depth_bottom_mm"] = [
        interval[1] for interval in intervals
    ]
    observations["is_t0"] = observations["timepoint"].str.upper().eq("T0")

    application_keys = ["site_id", "soil_group", "treatment"]
    application_dates = (
        observations.loc[observations["is_t0"]]
        .groupby(application_keys, sort=True, dropna=False)["sample_date"]
        .min()
        .rename("application_date")
    )
    observations = observations.join(application_dates, on=application_keys)
    if observations["application_date"].isna().any():
        raise ValueError(
            "At least one observation group has no T0 application date"
        )
    observations["days_since_application"] = (
        observations["sample_date"] - observations["application_date"]
    ).dt.days.astype(int)

    replicate_keys = [
        "site_id",
        "soil_group",
        "treatment",
        "herbicide",
        "depth_top_mm",
        "depth_bottom_mm",
        "sample_date",
    ]
    observations["replicate_id"] = (
        observations.groupby(
            replicate_keys,
            sort=True,
            dropna=False,
        ).cumcount()
        + 1
    )
    observations["is_non_detect"] = False
    observations["detection_limit_ug_kg"] = np.nan
    observations["is_zero_reported"] = (
        observations["concentration_ug_kg"] <= 0
    )
    non_detects = prepare_non_detects(
        observations["concentration_ug_kg"],
        observations["is_non_detect"],
        observations["detection_limit_ug_kg"],
    )
    observations = pd.concat(
        [observations.reset_index(drop=True), non_detects],
        axis=1,
    )
    observations["unit_status"] = "inferred_from_application_rate"

    sort_columns = [
        "site_id",
        "soil_group",
        "treatment",
        "herbicide",
        "sample_date",
        "depth_top_mm",
        "replicate_id",
    ]
    return observations.sort_values(
        sort_columns,
        kind="stable",
    ).reset_index(drop=True)


def infer_application_rate_g_ha(
    t0_concentration_ug_kg: ArrayLike,
    depth_top_mm: float,
    depth_bottom_mm: float,
    bulk_density_g_cm3: float,
) -> float:
    """Infer applied mass from positive top-layer T0 concentrations."""

    concentration = np.atleast_1d(
        np.asarray(t0_concentration_ug_kg, dtype=float)
    )
    if (
        len(concentration) == 0
        or not np.all(np.isfinite(concentration))
        or np.any(concentration <= 0)
    ):
        raise ValueError(
            "T0 concentrations must be finite and greater than zero"
        )
    soil_mass = soil_mass_kg_ha(
        depth_top_mm,
        depth_bottom_mm,
        bulk_density_g_cm3,
    )
    return float(math.exp(np.mean(np.log(concentration))) * soil_mass / 1e6)

File: R/R/observations.R
# Script: observations.R
# Objective: Prepare replicate-level herbicide observations for RCLT analysis.
# Author: Yi Yu
# Created: 2026-06-23
# Last updated: 2026-06-23
# Inputs: Herbicide workbooks or tidy concentration vectors.
# Outputs: Explicit depth intervals, non-detect fields, summaries, and inferred application rates.
# Usage: Use exported functions after library(rclt).
# Dependencies: readxl

#' Map workbook depth labels to sampling intervals
#'
#' @param sheet Workbook sheet or jurisdiction label.
#' @param depth_label Workbook depth label.
#' @return Numeric vector containing top and bottom depths in millimetres.
#' @export
depth_interval_mm <- function(sheet, depth_label) {
  key <- paste(
    toupper(trimws(sheet)),
    gsub("\\s+", "", tolower(trimws(depth_label))),
    sep = ":"
  )
  intervals <- list(
    "SA:10cm"  = c(0, 100),
    "SA:30cm"  = c(100, 300),
    "NSW:15cm" = c(0, 150),
    "NSW:30cm" = c(150, 300),
    "QLD:10cm" = c(0, 100),
    "QLD:30cm" = c(100, 300)
  )
  result <- intervals[[key]]
  if (is.null(result)) {
    stop("Unsupported sheet/depth combination: ", key, call. = FALSE)
  }
  result
}

#' Prepare observations with explicit non-detect handling
#'
#' @param concentration_ug_kg Reported concentrations.
#' @param is_non_detect Logical non-detect indicators.
#' @param detection_limit_ug_kg Detection limits in micrograms per kilogram.
#' @return Data frame containing analysis concentrations and handling flags.
#' @export
prepare_non_detects <- function(
  concentration_ug_kg,
  is_non_detect,
  detection_limit_ug_kg
) {
  lengths <- c(
    length(concentration_ug_kg),
    length(is_non_detect),
    length(detection_limit_ug_kg)
  )
  if (length(unique(lengths)) != 1L) {
    stop("Non-detect input vectors must have equal lengths.", call. = FALSE)
  }
  if (any(!is.finite(concentration_ug_kg))) {
    stop("Reported concentrations must be finite.", call. = FALSE)
  }
  if (any(is.na(is_non_detect))) {
    stop("is_non_detect cannot contain missing values.", call. = FALSE)
  }

  substituted <- is_non_detect &
    is.finite(detection_limit_ug_kg) &
    detection_limit_ug_kg > 0
  excluded_zero <- concentration_ug_kg <= 0 & !substituted
  analysis <- concentration_ug_kg
  analysis[substituted] <- detection_limit_ug_kg[substituted] / 2
  analysis[excluded_zero] <- NA_real_

  data.frame(
    analysis_concentration_ug_kg = analysis,
    lod_substituted              = substituted,
    excluded_zero                = excluded_zero
  )
}

#' Calculate grouped geometric concentration summaries
#'
#' @param data Data frame containing `analysis_concentration_ug_kg`.
#' @param group_columns Character vector of grouping columns.
#' @return Data frame with counts, geometric means, and geometric standard deviations.
#' @export
geometric_concentration <- function(data, group_columns) {
  if (!"analysis_concentration_ug_kg" %in% names(data)) {
    stop("data must contain analysis_concentration_ug_kg.", call. = FALSE)
  }
  if (!all(group_columns %in% names(data)) || length(group_columns) == 0L) {
    stop("group_columns must identify columns in data.", call. = FALSE)
  }

  key <- interaction(data[group_columns], drop = TRUE, lex.order = TRUE)
  split_data <- split(data, key)
  rows <- lapply(split_data, function(group) {
    values <- group$analysis_concentration_ug_kg
    values <- values[is.finite(values) & values > 0]
    first <- group[1, group_columns, drop = FALSE]
    first$n <- length(values)
    first$geometric_mean_ug_kg <- if (length(values)) {
      exp(mean(log(values)))
    } else {
      NA_real_
    }
    first$geometric_sd <- if (length(values) > 1L) {
      exp(stats::sd(log(values)))
    } else {
      NA_real_
    }
    first
  })
  result <- do.call(rbind, rows)
  rownames(result) <- NULL
  result
}

normalize_sample_date <- function(value) {
  if (inherits(value, "POSIXt") || inherits(value, "Date")) {
    return(as.Date(value))
  }
  if (is.numeric(value)) {
    return(as.Date(value, origin = "1899-12-30"))
  }
  as.Date(value)
}

column_or_default <- function(data, column, default) {
  if (column %in% names(data)) data[[column]] else rep(default, nrow(data))
}

prepare_workbook_sheet <- function(path, sheet) {
  raw <- as.data.frame(
    readxl::read_excel(path, sheet = sheet, .name_repair = "unique"),
    stringsAsFactors = FALSE
  )
  identifier_columns <- intersect(
    c(
      "Soil",
      "Irrigation",
      "Timepoint",
      "Crop_2024",
      "Depth",
      "Sample_date"
    ),
    names(raw)
  )
  herbicide_columns <- setdiff(names(raw), identifier_columns)
  site_id <- switch(
    toupper(sheet),
    SA  = "SA_Minnipa",
    NSW = "NSW_Griffith",
    QLD = "QLD_Wellcamp",
    stop("Unsupported workbook sheet: ", sheet, call. = FALSE)
  )

  frames <- lapply(herbicide_columns, function(herbicide) {
    concentration <- suppressWarnings(as.numeric(raw[[herbicide]]))
    keep <- is.finite(concentration)
    if (!any(keep)) {
      return(NULL)
    }
    data.frame(
      site_id             = site_id,
      source_sheet        = sheet,
      source_row          = which(keep) + 1L,
      soil_group          = trimws(as.character(raw$Soil[keep])),
      treatment           = trimws(as.character(
        column_or_default(raw, "Irrigation", "All")[keep]
      )),
      crop_2024           = trimws(as.character(
        column_or_default(raw, "Crop_2024", NA_character_)[keep]
      )),
      timepoint           = trimws(as.character(raw$Timepoint[keep])),
      depth_label         = trimws(as.character(raw$Depth[keep])),
      sample_date         = normalize_sample_date(raw$Sample_date[keep]),
      herbicide           = herbicide,
      concentration_ug_kg = concentration[keep],
      stringsAsFactors    = FALSE
    )
  })
  frames <- Filter(Negate(is.null), frames)
  if (!length(frames)) {
    stop("No herbicide concentrations found in sheet ", sheet, ".", call. = FALSE)
  }
  do.call(rbind, frames)
}

#' Read the herbicide dissipation workbook
#'
#' @param path Path to the source Excel workbook.
#' @param sheets Workbook sheets to import.
#' @return Replicate-level tidy observations with explicit depth intervals.
#' @export
read_herbicide_workbook <- function(
  path,
  sheets = c("SA", "NSW", "Qld")
) {
  if (!requireNamespace("readxl", quietly = TRUE)) {
    stop("Package 'readxl' is required to import workbooks.", call. = FALSE)
  }
  if (!file.exists(path)) {
    stop("Observation workbook does not exist: ", path, call. = FALSE)
  }

  observations <- do.call(
    rbind,
    lapply(sheets, function(sheet) prepare_workbook_sheet(path, sheet))
  )
  intervals <- t(vapply(
    seq_len(nrow(observations)),
    function(index) {
      depth_interval_mm(
        observations$source_sheet[index],
        observations$depth_label[index]
      )
    },
    numeric(2)
  ))
  observations$depth_top_mm <- intervals[, 1]
  observations$depth_bottom_mm <- intervals[, 2]
  observations$is_t0 <- toupper(observations$timepoint) == "T0"

  application_group <- interaction(
    observations[c("site_id", "soil_group", "treatment")],
    drop      = TRUE,
    lex.order = TRUE
  )
  application_dates <- tapply(
    as.numeric(observations$sample_date[observations$is_t0]),
    application_group[observations$is_t0],
    min
  )
  observations$application_date <- as.Date(
    unname(application_dates[as.character(application_group)]),
    origin = "1970-01-01"
  )
  if (any(is.na(observations$application_date))) {
    stop("At least one observation group has no T0 application date.", call. = FALSE)
  }
  observations$days_since_application <- as.integer(
    observations$sample_date - observations$application_date
  )

  replicate_group <- interaction(
    observations[
      c(
        "site_id",
        "soil_group",
        "treatment",
        "herbicide",
        "depth_top_mm",
        "depth_bottom_mm",
        "sample_date"
      )
    ],
    drop      = TRUE,
    lex.order = TRUE
  )
  observations$replicate_id <- stats::ave(
    seq_len(nrow(observations)),
    replicate_group,
    FUN = seq_along
  )
  observations$is_non_detect <- FALSE
  observations$detection_limit_ug_kg <- NA_real_
  observations$is_zero_reported <- observations$concentration_ug_kg <= 0
  non_detects <- prepare_non_detects(
    observations$concentration_ug_kg,
    observations$is_non_detect,
    observations$detection_limit_ug_kg
  )
  observations <- cbind(observations, non_detects)
  observations$unit_status <- "inferred_from_application_rate"

  observations[order(
    observations$site_id,
    observations$soil_group,
    observations$treatment,
    observations$herbicide,
    observations$sample_date,
    observations$depth_top_mm,
    observations$replicate_id
  ), ]
}

#' Infer applied mass from top-layer T0 concentration
#'
#' @param t0_concentration_ug_kg Positive T0 replicate concentrations.
#' @param depth_top_mm,depth_bottom_mm T0 layer bounds.
#' @param bulk_density_g_cm3 Whole-earth bulk density.
#' @return Inferred application rate in grams per hectare.
#' @export
infer_application_rate_g_ha <- function(
  t0_concentration_ug_kg,
  depth_top_mm,
  depth_bottom_mm,
  bulk_density_g_cm3
) {
  if (!length(t0_concentration_ug_kg) ||
      any(!is.finite(t0_concentration_ug_kg)) ||
      any(t0_concentration_ug_kg <= 0)) {
    stop("T0 concentrations must be finite and greater than zero.", call. = FALSE)
  }
  soil_mass <- soil_mass_kg_ha(
    depth_top_mm,
    depth_bottom_mm,
    bulk_density_g_cm3
  )
  exp(mean(log(t0_concentration_ug_kg))) * soil_mass / 1e6
}
